Its about time (#433)
quicwg/qlog: draft-ietf-quic-qlog-main-schema.md

## Timestamps {#time-based-fields}

Each event MUST include a "time" field to indicate the timestamp that it

> occurred. It is a duration measured from some point in time, in units of ???.

Reviewer: ```suggestion
occurred. It is a duration measured from some point in time; its units depend on the type of clock chosen and system used. Typically however, a millisecond resolution is employed.
```

Author: I don't understand this bit. If I have a float64 field, and I decide to use it to log "1.5" to mean one and a half seconds, but the logger tool thinks I meant one and a half milliseconds, then there's an interop issue. So I think we need to pick one form of "units of x with fractional Y", or add a field to articulate the concrete units and precision.

There are several options for generating and logging timestamps, these are
governed by the ReferenceTime type (optionally included in the "reference_time"
field contained in a trace's "common_fields" ({{common-fields}})) and TimeFormat
type (optionally included in the "time_format" field contained in the event
itself, or a trace's "common_fields").

There is no requirement that events in the same trace use the same time format.
However, using a single time format for related events can make them easier to
analyze.

> The reference time is defined as:

Reviewer: ```suggestion
The reference time governs from which point in time the "time" field values are measured and is defined as:
```

~~~ cddl
ReferenceTime = {
    clock_type: "system" / "monotonic" / text .default "system"
    epoch: RFC3339DateTime / "unknown" .default "1970-01-01T00:00:00.000Z"

    ? wall_clock_time: RFC3339DateTime
}

> RFC3339DateTime = text

Reviewer: ```suggestion
; RFC 3339 UTC string at millisecond resolution. i.e., YYYY-MM-DDTHH:mm:ss.sssZ
RFC3339DateTime = text
```

Author: I don't think we need this, its profiling the RFC even further.

~~~
{: #reference-time-def title="ReferenceTime definition"}

The required "clock_type" field represents the type of clock used for time
measurements. The value "system" represents a clock that uses system time,

> commonly measured against a chosen or well-known epoch. System time can jump
> forward or back. The value "monotonic" represents a clock that uses monotonic
> time, measured against an arbitrary and opaque epoch. Monotonic time is
> generally guaranteed to never go backwards.

Reviewer: ```suggestion
commonly measured against a chosen or well-known epoch. However, depending on the system, System time can potentially jump forward or back. In contrast, a clock using monotonic time is generally guaranteed to never go backwards. The value "monotonic" represents such a clock. 
```

The required "epoch" field is the start of the ReferenceTime. When using the
"system" clock type, the epoch field SHOULD have a date/time value using the
format defined in {{!RFC3339}}. However, the value "unknown" MAY be used.

When using the "monotonic" clock type, the epoch field MUST have the value
"unknown".

The optional "wall_clock_time" field can be used to provide an approximate

> date/time value that logging commenced if the epoch value is "unknown". It uses

Reviewer: ```suggestion
date/time value that logging commenced at if the epoch value is "unknown". It uses
```

the format defined in {{!RFC3339}}. Note that conversion of timestamps to
calendar time based on wall clock times cannot be safely relied on.

> The time format is defined as:

Reviewer: ```suggestion
The time format details how "time" values are encoded relative to the reference time and is defined as:
```

~~~ cddl
TimeFormat = "relative_to_epoch" /
             "relative_to_previous_event" .default "relative_to_epoch"
~~~
{: #time-format-def title="TimeFormat definition"}

relative_to_epoch:
: A duration relative to the ReferenceTime "epoch" field. This approach uses the
  largest amount of characters. It is good for stateless loggers. This is the default value of the "time_format" field.

relative_to_previous_event:
: A delta-encoded value, based on the previously logged value. The first event
  in a trace is always relative to the ReferenceTime. This approach uses the
  least amount of characters. It is suitable for stateful loggers.

Events in each individual trace SHOULD be logged in strictly ascending timestamp
order (though not necessarily absolute value, for the "relative_to_previous_event"
format). Tools MAY sort all events on the timestamp before processing them,
though are not required to (as this could impose a significant processing
overhead). This can be a problem especially for multi-threaded and/or streaming
loggers, who could consider using a separate post-processor to order qlog events
in time if a tool do not provide this feature.

> Tools SHOULD NOT assume the ability to derive the absolute timestamp of an event

Reviewer: ```suggestion
Tools SHOULD NOT assume the ability to derive the absolute calendar timestamp of an event
```

  },
  {
    "time": 1553986553577,
    "name": "quic:packet_received",
    "data": { ... },
  },
  {
    "time": 1553986553587,
    "name": "quic:packet_received",
    "data": { ... },
  },
  {
    "time": 1553986553597,
    "name": "quic:packet_received",
    "data": { ... },
  },
]
~~~
{: #rel-epoch-time-ex title="Relative to epoch timestamps"}

Example of a log using the relative_to_previous_event format:

~~~
"common_fields": {
    "time_format": "relative_to_previous_event",
    "reference_time": {
          "clock_type": "system",
          "epoch": "1970-01-01T00:00:00.000Z"
    },
},
"events": [
  {
    "time": 1553986553572,
    "name": "quic:packet_received",
    "data": { ... },
  },
  {
    "time": 5,
    "name": "quic:packet_received",
    "data": { ... },
  },
  {
    "time": 10,
    "name": "quic:packet_received",
    "data": { ... },
  },
  {
    "time": 10,
    "name": "quic:packet_received",
    "data": { ... },
  },
]
~~~
{: #rel-last-event-time-ex title="Relative-to-previous-event timestamps"}

Example of a monotonic log using the relative_to_epoch format:

~~~
"common_fields": {
    "time_format": "relative_to_epoch",
    "reference_time": {

>           "clock_type": monotonic,

Reviewer: ```suggestion
          "clock_type": "monotonic",
```

{
    "events": [{
            "group_id": "127ecc830d98f9d54a42c4f0842aa87e181a",
            "protocol_type": ["QUIC","HTTP/3"],
            "time_format": "relative_to_epoch",
            "reference_time": {
              "clock_type": "system",

>               "epoch": "2019-03-39T:22:55:53.572Z"

Reviewer: ```suggestion
              "epoch": "2019-03-29T:22:55:53.572Z"
```

            },

            "time": 2,
            "name": "quic:packet_received",
            "data": { ... }
        },{
            "group_id": "127ecc830d98f9d54a42c4f0842aa87e181a",
            "protocol_type": ["QUIC","HTTP/3"],
            "time_format": "relative_to_epoch",
            "reference_time": {
              "clock_type": "system",

>               "epoch": "2019-03-39T:22:55:53.572Z"

Reviewer: ```suggestion
              "epoch": "2019-03-29T:22:55:53.572Z"
```

{
    "common_fields": {
        "group_id": "127ecc830d98f9d54a42c4f0842aa87e181a",
        "protocol_type": ["QUIC","HTTP/3"],
        "time_format": "relative_to_epoch",
        "reference_time": {
            "clock_type": "system",

>             "epoch": "2019-03-39T:22:55:53.572Z"

Reviewer: ```suggestion
            "epoch": "2019-03-29T:22:55:53.572Z"
```
